### ISSUE TITLE
[docs] Idea collection of examples of custom ops / inline torch extensions

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#106667]
### 🚀 The feature, motivation and pitch

Following up on https://github.com/pytorch/pytorch/issues/68066#issuecomment-1611836317:

<hr>
- a complete example of a custom op (maybe with forward/backward and registration) implemented via load_inline with only CUDA op or both CPU/SSE (how to properly do SSE impls?) and CUDA code - maybe for bilateral filter or batched base64 encoding/decoding or some bit magics like popcnt (https://github.com/pytorch/pytorch/issues/32867 or groupbitsum from https://github.com/Felix-Petersen/difflogic)

- maybe general discussion of whether `cupy` api/features for this are worth being compatible with? pure nvrtc compilation? pure C compilation / marshalling (of pure pointers + numbers + C strings + ctype structs?)?

- support for launcher-less kernels via `cupy`-like api? (also a good fit for nvrtc) for instance, here https://github.com/pytorch/pytorch/blob/main/test/test_cpp_extensions_jit.py#L338 there's even an unintuitive need for a c++ source for function declaration. it would be nicer to do without and be able to directly call the kernel

- example of calling a C function consuming raw data pointers or DLPack structures. It often is much simpler to have these kinds of interfaces for thin C wrappers (e.g. audio decoding). It would be nice to have this example as inline, as often it's just dozen lines of C++ code in extern C function to call some underlying library like BLAS / cuBLAS / some media encoding-decoding.

- example of using extra functions cuBLAS / cudnn / blas. Maybe an example here could be some code using `ctypes`? it should be explained how to correctly marshal CUDA pointers.

<hr>

More ideas in https://github.com/pytorch/pytorch/pull/105947#issuecomment-1655693163:

- also interesting would be examples on how to register ops currently missing in pytorch, e.g. adding a quantized impl or some missing sparse op, so that dispatcher would use it and let go the exception that some op impl can't be found

- also interesting an example of adding a custom backward impl via a decorator

- also interesting an example of cuda op, but without having to manually add a C++ launcher stub (e.g. so that pytorch would generate the kernel launcher and maybe that the kernel can be compiled by nvrtc - like in CuPy)

- and also the CuPy-like API could be worth a discussion

- also interesting would be an example of adding a pure C-op passing/consuming a dlpack tensor or even pure pointers - useful for interaction with C libraries of image/audio encoding/decoding

**Comment by albanD** [comment]
I would be very happy to help reviewing such a PR if you're interested!

**Comment by vadimkantorov** [comment]
I would say, it'd be a huge PR as my ideas above are not very structured yet. I would propose I keep on making descriptions of concrete examples and then we can factor out them to separate issues

**Comment by vadimkantorov** [comment]
It would be good to have an example of launching kernels from FasterTransformer or OneFlow without compiling/installing these packages (mainly dealing with marshalling). This would allow simpler benchmarks comparing perf without having to add more pip dependencies

**Comment by vadimkantorov** [comment]
It would be nice to have example of directly using CUDA kernel source files from [FasterTransformer](https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/kernels/layernorm_kernels.cu ) or [OneFlow](https://github.com/Oneflow-Inc/oneflow/blob/2d24fe08be1b1bedcc22fb409c5d688924ce89fc/oneflow/user/kernels/layer_norm_gpu_kernel.cu). Somehow getting a compiled kernel without fully installing a python Package gives a nice feeling of experimentation and flexibility :)

E.g. it would be nice to include FasterTransformer as submodule and just have a `my_kernels = load('fused_layer_norm.cu')` in user's PyTorch code (without having to properly build/install 
 the full FasterTransformer) - as in CuPy.

**Comment by vadimkantorov** [comment]
It appears that FasterTransformer development is now stopped in benefit of TransformerEngine, so might be good of having some tools of launching individual FasterTransformer kernels in PyTorch without having to install/build the full FasterTransformer

**Comment by vadimkantorov** [comment]
@albanD which points of "distilled ideas" from the header post are ready to be factored out as separate issues?

**Comment by mikekgfb** [comment]
If you're looking at https://github.com/pytorch/pytorch/pull/117500, that's all you need to write small fragments to "hook up" opencv or other libraries with basically inline C/C++.  Have you experimented with this yet?  What works, what does not?

I'll be curious to understand what we can achieve with this - as I've also looked an simplified integration of C/C++ snippets.  Let's focus on C/C++ for now (plus we have a separate triton solution). 

Understood that if this works well, we might need other compilers, including for CUDA, ROCM, and other backends. Once we have a proof of concept and can show utility, adding other toolchains should be pretty mechanical.

**Comment by vadimkantorov** [comment]
no, at some point I experimented with: 1) inline C++ extensions, 2) C-library (not dependent on libtorch) + DLpack for marshalling + ctypes

Both of these work, but it would be nice to have examples of such extensions in the docs + maybe improved UX (CuPy-style) for allowing

+ one discussible thing is that I personally consider a good thing if such "extension" does not depend on libtorch and can accept only C-objects like DLPack, the fewer dependencies - the bettter + also if it's just C code, it compiles faster (of course if it does more work with torch, then it's useful to link it to libtorch to allow some tensor processing))

**Comment by vadimkantorov** [comment]
Regarding OpenCV integration example, I would also expect that such an example clarifies a relation between libopencv-dev's built-in OpenMP or some other multithreading and the multithreading built-in PyTorch. As I imagine that using different impls of openmp might backfire (e.g. known to have problems in calling OpenCV in DataLoader threads)

**Comment by vadimkantorov** [comment]
Also, for a good, simple interop story, python's from_blob/frombuffer API would need to be polished: https://github.com/pytorch/pytorch/issues/107112#issuecomment-1891515020

**Comment by mikekgfb** [comment]
I agree that this needs some investigation. If you’re passionate about it, the best approach may be to figure out OpenCV interopability with respect to openmap.

If they’re linked into a single binary, then they may need to have the same version because there ‘d be a symbol clash during linking.

This is best driven by a domain expert being familiar with open cv and PyTorch.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
